lib/gc/Transforms/MemRefToCPURuntime.cpp

```diff
@@ -54,48 +54,49 @@ struct AlignedAllocLowering : public OpRewritePattern<memref::AllocOp> {
     return success();
   }
 };
-
-struct AlignedDeallocLowering : public OpRewritePattern<memref::DeallocOp> {
-  using OpRewritePattern<memref::DeallocOp>::OpRewritePattern;
-  LogicalResult matchAndRewrite(memref::DeallocOp op,
-                                PatternRewriter &rewriter) const final {
-    auto loc = op->getLoc();
-    Value memref = op.getMemref();
-    cpuruntime::DeallocOp newDeallocOp =
-        rewriter.create<cpuruntime::DeallocOp>(loc, memref);
-    if (hasParallelParent(op))
-      newDeallocOp.setThreadLocal(true);
-    rewriter.eraseOp(op);
-    return success();
-  }
-};
-
 struct ConvertMemRefToCPURuntime
     : public impl::ConvertMemRefToCPURuntimeBase<ConvertMemRefToCPURuntime> {
 
   void runOnOperation() final {
     auto *ctx = &getContext();
-    // Create a local set to store operations that should not be transformed.
-    llvm::SmallSet<Operation *, 16> noTransformOps;
 
-    // Walk through the module to find func::FuncOp instances.
+    // Create deallocOp accoresponding to the alloca's localtion
     getOperation()->walk([&](func::FuncOp funcOp) {
-      BufferViewFlowAnalysis analysis(funcOp);
-      // Now walk through the operations within the func::FuncOp.
-      funcOp.walk([&](Operation *op) {
-        if (op->hasTrait<OpTrait::ReturnLike>()) {
-          for (Value operand : op->getOperands()) {
-            if (isa<MemRefType>(operand.getType())) {
-              auto aliases = analysis.resolveReverse(operand);
-              // Check if any of the returned memref is allocated within scope.
-              for (auto &&alias : aliases) {
-                if (Operation *allocOp =
-                        alias.getDefiningOp<memref::AllocOp>()) {
-                  noTransformOps.insert(allocOp);
-                }
-              }
+      funcOp.walk([&](memref::AllocaOp op) {
+        Region *parentRegion = op->getParentRegion();
+        OpBuilder builder(op);
+        // Find the first deallocOp in the current region
+        cpuruntime::DeallocOp firstDeallocOp;
+        for (Block &block : parentRegion->getBlocks()) {
+          for (Operation &operation : block) {
+            if (auto deallocOp = dyn_cast<cpuruntime::DeallocOp>(&operation)) {
+              firstDeallocOp = deallocOp;
+              break;
             }
           }
+          if (firstDeallocOp)
+            break;
+        }
+
+        // If a deallocOp was found, insert the new dealloc before it
+        if (firstDeallocOp) {
+          builder.setInsertionPoint(firstDeallocOp);
+        } else {
+          // If no deallocOp was found, insert at the end of the region before
```

Comment:
Please insert deallocop at the end of the parent alloca_scope op. It may not be the same as `op->getParentRegion()`

And I am not sure why we need to find `firstDeallocOp`?

Another issue is that, if the alloca size is very small, I think we'd better keep using the alloca instead of a runtime function. In GC v1 I remember that we have a size limit at ~128 bytes.

Reply:
> Another issue is that, if the alloca size is very small, I think we'd better keep using the alloca instead of a runtime function. In GC v1 I remember that we have a size limit at ~128 bytes.

This is what I concern about, the current pass is trying to convert all the alloca (no matter the size) into runtime allocator which is not that efficient.

Reply:
Thanks for the suggestion. 

1. use `getParentWithTrait<OpTrait::AutomaticAllocationScope>()` to locate the dealloc op's position.
2. need to use a while loop to find the `firstDeallocOp` to ensure the deallocation order obeys FILO.
```
// Move the insertion point back if the operation before is a deallocOp
      Operation *currentOp = builder.getInsertionPoint()->getPrevNode();
      while (currentOp && isa<cpuruntime::DeallocOp>(currentOp)) {
        // Move the insertion point before the found deallocOp
        builder.setInsertionPoint(currentOp);
        currentOp = currentOp->getPrevNode();
      }
```
3. limit size is added.

Reply:
> need to use a while loop to find the firstDeallocOp to ensure the deallocation order obeys FILO.

Please double check the IR result of `promote-allocation-to-stack`. AFAIK, there is no dealloc for alloca ops - it is "deallocated" via the end `AutomaticAllocationScope`.

Reply:
Yes, from `promote-allocation-to-stack` to the end of `deallocation pipeline`, no dealloc will be created for alloca ops. It will be dealloced automatically at the end of `AutomaticAllocationScope`.

The above `while loop` only searches `cpuruntime::deallocOp`, to make sure the generated dealloc order as FILO.

Reply:
why do we need to check cpuruntime::deallocOp that has just been created by us in this pass?

Reply:
When we walk through the `AutomaticAllocationScope`. If we get the below IR, the first allocated buffer `m0` will be first deallocated at the end of the scope. Then the next allocated buffer `m1` will be the next op to create the `dealloc` op, whose position should be the previous node of `dealloc %m0`
```
func.func @alloca_sequence() {
  // CHECK-LABEL: func @alloca_sequence()
  // CHECK: %[[m0:.*]] = cpuruntime.alloc() : memref<128xf32>
  // CHECK: %[[m1:.*]] = cpuruntime.alloc() : memref<128xf32>
  // CHECK: %[[m2:.*]] = cpuruntime.alloc() : memref<128xf32>
  %alloc = memref.alloca() : memref<128xf32>
  %alloc_0 = memref.alloca() : memref<128xf32>
  %alloc_1 = memref.alloca() : memref<128xf32>
  // CHECK: cpuruntime.dealloc %[[m2]] : memref<128xf32>
  // CHECK: cpuruntime.dealloc %[[m1]] : memref<128xf32>
  // CHECK: cpuruntime.dealloc %[[m0]] : memref<128xf32>
  return
}
```

Reply:
traverse the `allocaOp` first, and then insert `deallocOp` in a reverse order now.